kong/pdk/log.lua

```diff
@@ -465,11 +474,172 @@ local _log_mt = {
 }
 
 
+local function get_default_serialize_values()
+  if ngx.config.subsystem == "http" then
+    return {
+      { key = "request.headers.authorization", value = "REDACTED", mode = "replace" },
+      { key = "request.headers.proxy-authorization", value = "REDACTED", mode = "replace" },
+    }
+  end
+
+  return {}
+end
+
+---
+-- Sets a value to be used on the `serialize` custom table
+--
+-- Logging plugins use the output of `kong.log.serialize()` as a base for their logs.
+--
+-- This function allows customizing such output.
+--
+-- It can be used to replace existing values on the output.
+-- It can be used to delete existing values by passing `nil`.
+--
+--
+-- @function kong.log.set_serialize_value
+-- @phases certificate, rewrite, access, header_filter, body_filter, log
+-- @tparam string key the name of the field.
+-- @tparam number|string|boolean|table value value to be set. When a table is used, its keys must be numbers, strings, booleans, and its values can be numbers, strings or other tables like itself, recursively.
+-- @tparam table options can contain two entries: options.mode can be `set` (the default, always sets), `add` (only add if entry does not already exist) and `replace` (only change value if it already exists).
+-- @treturn table the request information table
+-- @usage
+-- -- Adds a new value to the serialized table
+-- kong.log.set_serialize_value("my_new_value", 1)
+-- assert(kong.log.serialize().my_new_value == 1)
+--
+-- -- Value can be a table
+-- kong.log.set_serialize_value("my", { new = { value = 2 } })
+-- assert(kong.log.serialize().my.new.value == 2)
+--
+-- -- It is possible to change an existing serialized value
+-- kong.log.set_serialize_value("my_new_value", 3)
+-- assert(kong.log.serialize().my_new_value == 3)
+--
+-- -- Unset an existing value by setting it to nil
+-- kong.log.set_serialize_value("my_new_value", nil)
+-- assert(kong.log.serialize().my_new_value == nil)
+--
+-- -- Dots in the key are interpreted as table accesses
+-- kong.log.set_serialize_value("my.new.value", 4)
+-- assert(kong.log.serialize().my.new_value == 4)
+--
+local set_serialize_value
+
+do
+
+  local function is_valid_value(v)
+    local t = type(v)
+    if v == nil or t == "number" or t == "string" or t == "boolean" then
+      return true
+    end
+
+    if t ~= "table" then
+      return false
+    end
+
+    for k, val in pairs(v) do
+      t = type(k)
+      if (t ~= "string"
+          and t ~= "number"
+          and t ~= "boolean")
+        or not is_valid_value(val)
+      then
+        return false
+      end
+    end
+
+    return true
+  end
+
+  set_serialize_value = function(key, value, options)
```

Comment:
little known fact: the "normal" `function name(args)...` form can assign to an upvalue just as well:
```lua
local visible_function
do
  local private_values
  ...
  function visible_function()
    ...
  end
end
```
works as expected :smiley: !

Reply:
It is my personal preference. This line:

```
function visible_function()
```

Triggers my mental `YOU ARE SETTING A GLOBAL VARIABLE`-alarm, so every time I see that I must do a double-take and make sure that the variable is `local`-ized somewhere. Even if I have luacheck, which takes care of that sort of thing, it is still in the background eating cycles until I manually do the check.

The assignment syntax does not trigger that alarm for me so it is "cheaper to read" for me 😁.